### PR TITLE
Update dev container for swift

### PR DESCRIPTION
**Pull request by damuellen**
Changing the setting and installed extension. To reflect the latest state of Swift development.
[introducing-swift-for-visual-studio-code](https://forums.swift.org/t/introducing-swift-for-visual-studio-code/54246)

### CODE REVIEW

**Comment by adam-fowler** [comment]
There are quite a lot of changes in here. Not sure we need to include all of the extensions in here. Plus I think it is  worthwhile bringing in @vknabel to discuss if there should be a transition from the original swift extension or should we add a new swift devcontainer for the SSWG extension. 

I think this is also a good time to discuss what is actually needed in the devcontainer. For me there appears to be a lot of unnecessary stuff here. Why are we installing node, python? Is this something MS expects?

Including @0xTim

**Comment by adam-fowler** [comment]
We can remove the lldb install for sure though, it is overriding the swift version of lldb.

**Comment by 0xTim** [comment]
I'm guessing node is needed to run the VSCode environment. Python was probably installed for LLDB or to install Swift, but that's now been removed.

As for all the extra extensions - I agree that we're installing a lot we probably don't need. Sticking to just `"sswg.swift-lang"` as a default would be my preferred option (I have no idea if we need to list dependencies like LLDB however)

**Comment by adam-fowler** [comment]
You don't need node.js well at least I can build a devcontainer which doesn't install it. Could it be included in the swift docker images already?

File: containers/swift/.devcontainer/devcontainer.json [hunk 1]
```diff
@@ -17,23 +17,27 @@
 
 	// Set *default* container specific settings.json values on container create.
 	"settings": {
-		"lldb.adapterType": "bundled",
-		"lldb.executable": "/usr/bin/lldb",
-		"sde.languageservermode": "sourcekite",
-		"swift.path.sourcekite": "/usr/local/bin/sourcekite"
+		"lldb.library": "/usr/lib/liblldb.so",
+		"sourcekit-lsp.toolchainPath": "/usr/bin/",
```
Comment (adam-fowler) [inline containers/swift/.devcontainer/devcontainer.json:21]: You shouldn't need to setup sourcekit-lsp paths

File: containers/swift/.devcontainer/Dockerfile [hunk 2]
```diff
@@ -14,15 +14,8 @@ ARG USER_GID=$USER_UID
 COPY library-scripts/common-debian.sh /tmp/library-scripts/
 RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
     && /bin/bash /tmp/library-scripts/common-debian.sh "${INSTALL_ZSH}" "${USERNAME}" "${USER_UID}" "${USER_GID}" "${UPGRADE_PACKAGES}" "true" "true" \
-    && apt-get -y install --no-install-recommends lldb python3-minimal libpython3.7 \
```
Comment (adam-fowler) [inline containers/swift/.devcontainer/Dockerfile:17]: 👍 swift docker images install a later version of python.

**Review by adam-fowler — COMMENTED** [review]
As I've said in previous comments, can we keep this to just the `sswg.swift-lang` extension. Plus we need to work out what we want to do with the other swift vscode extension.

**Comment by vknabel** [comment]
Hi, thanks for the mention. I haven't found time for a specific migration plan, yet. But I do also vote for `sswg.swift-lang` without msde.

The few benefits msde still has, are mostly restricted to very old Swift versions and other niche use cases. Both user groups won't use this dev container.

Regarding the other extensions I'd also prefer using `sswg.swift-lang` only. Removing extensions or features is hard and we will never have such a great opportunity to do so.

In case we wouldn't, we'd need set up and install all formatter extensions (for apple's and nicklockwood's), not just one to stay neutral and not opinionated. And in that case also Swiftlint as it's widespread. IMHO far too much extensions.

_(Do I remember correctly, `sswg.swift-lang` also has a formatter built in?)_

Regarding the other extensions: haven't heard of them until now. But to be fair, haven't been active for a while.

**Comment by damuellen** [comment]
I removed the other extensions, even though they are all quite useful.

File: containers/swift/.devcontainer/devcontainer.json [hunk 3]
```diff
@@ -17,23 +17,20 @@
...
-		"swift.path.sourcekite": "/usr/local/bin/sourcekite"
+		"lldb.library": "/usr/lib/liblldb.so",
 	},
 
 	// Add the IDs of extensions you want installed when the container is created.
 	"extensions": [
-		"vknabel.vscode-swift-development-environment",
-		"vadimcn.vscode-lldb"
+		"sswg.swift-lang",
+		"vadimcn.vscode-lldb",
```
Comment (0xTim) [inline containers/swift/.devcontainer/devcontainer.json:26]: This is not required as it's pulled in automatically by `sswg.swift-lang`

**Review by 0xTim — COMMENTED** [review]
LGTM other than the LLDB requirement as it's pulled in automatically

**Comment by 0xTim** [comment]
@tomerd I wonder if there's anyone we can poke about getting this added as an official language rather than a community one

File: containers/swift/.devcontainer/devcontainer.json [hunk 4]
```diff
@@ -16,9 +16,7 @@
 	],
 
 	// Set *default* container specific settings.json values on container create.
-	"settings": {
-		"lldb.library": "/usr/lib/liblldb.so",
```
Comment (0xTim) [inline containers/swift/.devcontainer/devcontainer.json]: @damuellen sorry I may have confused things - this is definitely still required - it's the `"vadimcn.vscode-lldb",` line in extensions we don't need

**Comment by adam-fowler** [comment]
LGTM

@0xTim Regarding Official vs Community. We can just edit the .devcontainer to say "Swift" instead of "Swift (community)". I'm not sure if that's affects these changes getting merged.

**Comment by Chuxel** [comment]
@adam-fowler The "Community" indicator is more about whether it's a 1st party maintained definition or not. More of a statement around support and ongoing maintenence than validity.  

Otherwise, this PR looks great!  Happy to merge if you feel like it is ready.

**Comment by damuellen** [comment]
@Chuxel Should I open a new PR with squashed commits ?

**Comment by Chuxel** [comment]
@damuellen Nope, no need - I'll do that at merge time (via "Squash and merge").

@adam-fowler In concept, we could have it say `(Swift Server Working Group)`. This appears in the "Add Dev Container Configuration Files" UX.  

image

The README should also get updated with the go-forward owner conacts we can "@" in the event an issue or PR is raised.  Right now the README says @cloudnull.

**Comment by 0xTim** [comment]
@Chuxel if we can have just `Swift` in the name (without community) that would be great. Understand if that's not possible yet unless we can find an MS staff to step up. 

In terms of the README @damuellen if you want to update that for Swift, I'm happy to be listed as a maintainer. It doesn't look like the SSWG team can be tagged so you can list me, @adam-fowler if he's happy, yourself if you want and @cloudnull if he still wants to be tagged

**Comment by damuellen** [comment]
@0xTim Sure I can do this. But I will still wait for the feedback of others.

**Comment by adam-fowler** [comment]
Yeah you can put me on there

**Comment by Chuxel** [comment]
@0xTim Yeah this is more about signaling who is maintaining and keeping the images up to date, etc. The implicit assumption ends up being the VS Code team.

I'd be happy to change this to something indicating the organization maintaining the image rather than `(Community)`, however. e.g. `(Official Swift Working Group Definition)`

Otherwise, happy to merge.

**Comment by tomerd** [comment]
@adam-fowler @0xTim lets discuss and finalize on our next meeting

**Comment by 0xTim** [comment]
@Chuxel this is good to merge then. We can submit a future PR with clarified naming once that's been worked out

**Review by cloudnull — COMMENTED** [review]
LGTM 👍🏼

**Comment by Chuxel** [comment]
Sounds great! Merging. Thanks for the contribution!!

**Comment by bamurtaugh** [comment]
Hi @adam-fowler, @0xTim, and @cloudnull 👋!

Our team is working on the [open dev container spec](https://containers.dev/) to make dev containers suitable for any user in any tool. As part of this, we now have a [`devcontainers` GitHub org](https://github.com/devcontainers), and we place repos related to the spec in this org.
 We'll be hosting a specific set of Features, images, and templates that were previously in this vscode-dev-containers repository in repos in that org instead.

We want enable teams to make updates and fixes to their dev container resources at any cadence they like, rather than requiring the review and merge process required in this repo. Thus, we now have ways to self-publish [Features](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features-distribution.md) and [templates](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-templates-distribution.md). There's more info in [this issue](https://github.com/microsoft/vscode-dev-containers/issues/1589) too.

It'd be fantastic if you were interested in self-publishing the Swift template in your own repo, following the process linked above. We're more than happy to provide help and feedback along the way, and your feedback about the process is a huge help to our team too.

Let me know if you have any questions, and we're looking forward to hearing your thoughts. Thank you!

**Comment by adam-fowler** [comment]
Hi @bamurtaugh 

I was just thinking I need to update the swift container to include the latest swift version, so this is well timed. Yes managing the swift devcontainer template ourselves would be great. I would be happy to self-publish the template. As this would most likely be hosted in the swift-server organisation I will bring this up at the next swift server workgroup meeting. In the meantime I will look through the documents you have linked to.

**Comment by bamurtaugh** [comment]
That's great news, thanks @adam-fowler! Looking forward to hearing feedback from you and the working group.

**Comment by adam-fowler** [comment]
Hi @bamurtaugh it looks like we will be going forward with hosting our own versions of the swift devcontainer templates. I am away for the next 10 days so will start looking into it once I get back.

**Comment by bamurtaugh** [comment]
Fantastic, thanks for letting us know, @adam-fowler!

**Comment by adam-fowler** [comment]
Hi @bamurtaugh ,
We are going to host our own version of the swift dev container template. I have setup a repo https://github.com/swift-server/vscode-swift-devcontainer. If you could have a look at it to verify everything is ok that'd be great.

The template at the moment is pretty much a copy of the one in vscode-dev-containers. It currently has a number of options (install zsh, update packages, install NodeJS) which we may want to change. Are these options expected to be available or do we have free reign to change as we see fit?

**Comment by bamurtaugh** [comment]
Hi @adam-fowler! Thanks so much for sharing this. The template looks good to me. I'm also sharing it in our new community Slack channel for other eyes. We'd love to have you join our community Slack by the way - here's where you can comment if you'd like to join: https://github.com/orgs/devcontainers/discussions/3.

> Are these options expected to be available or do we have free reign to change as we see fit?
The contents of the template are totally up to your team - please feel free to adjust it as you see fit. As it sounds like these could be breaking changes for users currently using the template, we might recommend putting info in the REAMDE and/or a pinned issue in your repo to notify users about changes.

**Comment by adam-fowler** [comment]
> The contents of the template are totally up to your team - please feel free to adjust it as you see fit. As it sounds like these could be breaking changes for users currently using the template, we might recommend putting info in the REAMDE and/or a pinned issue in your repo to notify users about changes.

Any changes we would make would be heavily publicised. Depending on responses we could also create a legacy container which was a copy of the original.

How do I go about publishing this template and will it clash with the currently published one?

**Comment by Chuxel** [comment]
@adam-fowler The starter template has some info on this - https://github.com/devcontainers/template-starter#distributing-templates. You can use a GitHub Action to create the needed OCI Artifact - just copy the `.github` folder into your repo and you'll have a release workflow you can manually trigger. You can update the triggers based on how you want to manage things as well. Setup is completely automatic. You can then make the artifact public (info in the repo README) so anyone can download it.  

You can do a PR to add your template collection to the index so it shows up in UX - this is one time setup for the repo, so you can add as a many template vacations as you want at that point and they'll all show up.

We can then remove the one here from the index just by deleting the devcontainer-template.json file so there's no conflict.

**Comment by adam-fowler** [comment]
@Chuxel Unfortunately Apple won't allow the usage of the GH action. So I need to do the OCI Artifact generation and release manually

**Comment by bamurtaugh** [comment]
Hey @adam-fowler, happy new year! I just wanted to check in - please let us know if you needed any help with releasing your Template, we're happy to help.

**Comment by adam-fowler** [comment]
Hi, I just haven't got around to it. I'll see if I can find time this week.

**Comment by adam-fowler** [comment]
@bamurtaugh having looked through all the documentation and played with the devcontainer cli tool can I just use 
```
devcontainer template publish <path to folder containing devcontainer-template.json>
```

**Comment by bamurtaugh** [comment]
@adam-fowler, the `templates publish` command is a good choice!

Which container registry are you going to use?

**Comment by adam-fowler** [comment]
@bamurtaugh I hadn't really thought about that. I assume we would use the ghcr.io one. I can verify that.

**Comment by adam-fowler** [comment]
@bamurtaugh can we use hub.docker.com?

**Comment by bamurtaugh** [comment]
Yes! You can distribute using any OCI registry that implements the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec): https://containers.dev/implementors/templates-distribution/#oci-registry.

**Comment by adam-fowler** [comment]
Hi, we're back to publishing in ghcr.io. 

First I logged into ghrc.io

```
echo $CR_PAT | docker login ghcr.io -u adam-fowler --password-stdin
```

Then I tried to publish the template using

```
devcontainer templates publish -r ghcr.io -n swift-server/swift-dev-container ./src
```

But am getting the following error

```
[2 ms] @devcontainers/cli 0.27.1. Node.js v16.15.0. darwin 22.1.0 arm64.
[2 ms] Packaging template collection...
[3 ms] Processing template: swift...
[14 ms] Packaged 1 templates!
[14 ms] Processing template: swift...
[15 ms] Fetching published versions...
[905 ms] [httpOci] Could not fetch bearer token for 'ghcr.io': Error: HTTP 403: Forbidden
[906 ms] [httpOci] ERR: Failed to fetch Bearer token from registry.
[906 ms] Request failed
[906 ms] (!) ERR: Failed to publish 'ghcr.io/swift-server/swift-dev-container/swift'
```
I have also tried namespace `adam-fowler/swift-dev-container` and that doesn't work either.

**Comment by samruddhikhandale** [comment]
Hi @adam-fowler 👋 

> devcontainer templates publish -r ghcr.io -n swift-server/swift-dev-container ./src

Would it be possible to append `--log-level trace` to the above command ☝️  and paste the output here? thanks!

**Comment by adam-fowler** [comment]
Here you go
```
[1 ms] @devcontainers/cli 0.27.1. Node.js v16.15.0. darwin 22.1.0 arm64.
[1 ms] Packaging template collection...
[2 ms] Processing template: swift...
[12 ms] Packaged 1 templates!
[12 ms] Processing template: swift...
[12 ms] > input: ghcr.io/swift-server/swift-dev-container/swift
[12 ms] >
[12 ms] > resource: ghcr.io/swift-server/swift-dev-container/swift
[12 ms] > id: swift
[12 ms] > version: latest
[12 ms] > owner: swift-server
[12 ms] > namespace: swift-server/swift-dev-container
[12 ms] > registry: ghcr.io
[12 ms] > path: swift-server/swift-dev-container/swift
[12 ms] Fetching published versions...
[492 ms] [httpOci] Attempting to authenticate via 'Bearer' auth.
[495 ms] [httpOci] No authentication credentials found for registry 'ghcr.io'. Accessing anonymously.
[495 ms] [httpOci] Attempting to fetch bearer token from:  https://ghcr.io/token?service=ghcr.io&scope=repository:swift-server/swift-dev-container/swift:pull
[891 ms] [-] HTTP request failed with status code 403: : Forbidden
[892 ms] [httpOci] Could not fetch bearer token for 'ghcr.io': Error: HTTP 403: Forbidden
[892 ms] [httpOci] ERR: Failed to fetch Bearer token from registry.
[892 ms] Request failed
[892 ms] (!) ERR: Failed to publish 'ghcr.io/swift-server/swift-dev-container/swift'
```

**Comment by joshspicer** [comment]
Right now there are several ways to provide registry credentials to the CLI.  Unfortunately we don't yet have support for reading from OS credential helpers (like what is probably being used your Mac). 

Here's the code that looks for the credentials:

https://github.com/devcontainers/cli/blob/d02379c581b55233946f407a09a8ee1eeeff5fb4/src/spec-configuration/httpOCIRegistry.ts#L131-L164

The quickest way for you to get publishing from your local machine should be to do:
```
GITHUB_TOKEN="$CR_PAT" devcontainer templates publish -r ghcr.io -n swift-server/swift-dev-container ./src
```

**Comment by joshspicer** [comment]
Awesome! Making a note to document this more clearly today.

**Comment by adam-fowler** [comment]
@bamurtaugh Here is the PR for adding the Swift dev container https://github.com/devcontainers/devcontainers.github.io/pull/150. I hope it all works.

At the moment it looks like the old devcontainer, but we hope to update to the newer devcontainer feature style that all the new containers seem to have moved to.

**Comment by samruddhikhandale** [comment]
Thanks, looks good to me! ✨ 

image

Have https://github.com/microsoft/vscode-dev-containers/pull/1713 opened to deprecate the older Template.

**Comment by adam-fowler** [comment]
Nothing happens when I select either of these. I'm currently on a really bad internet connection so don't know if that is the cause.

**Comment by bamurtaugh** [comment]
Thanks so much for the PR, @adam-fowler! This is really exciting, and we really appreciate your collaboration.

I'm able to add the Swift Template successfully in the Dev Containers extension. If you're able to 1) try again as you get a more stable internet connection, 2) let us know if Swift is the only Template this is happening for (suggesting there is something unique there for some reason), that'd be great.

**Comment by 0xTim** [comment]
I've tested this and it's working for me so I think we're all good!

**Comment by adam-fowler** [comment]
Looks like it's working now

**Pull request merged**